### PR TITLE
Refactor async code synchronization

### PR DESCRIPTION
**Pull request by ansasaki**
Use messages for synchronization instead of mutexes and condvar.  Add
separate workers for keys, payload, and revocation execution, as well as
a worker to handle the revocations coming from ZeroMQ channel.

The workers  hold the state locally instead of storing all data in server
context (QuoteData).

The keys worker holds the received U and K keys and is responsible for
combining them to obtain K.  It also provides the K key when requested
by the /keys/verify endpoint to prove that the resulting K is correct.

When the U and V keys combination is successful (the HMAC verification
is successful), the keys worker wakes the payloads worker sending the
obtained K key and the encrypted payload.

The payloads worker is responsible for decrypting and running the
payload sent by the keys worker.

After decrypting and running the payload, the payloads worker will send
messages to the revocation worker and ZMQ worker to start listening for
revocations.

This also implements graceful shutdown on Ctrl + C (SIGINT) signal.
A dedicated task is spawned to receive the signal and stop the other workers.

Unit tests were implemented for the changes.

Fixes: #363, #447

### CODE REVIEW

File: keylime-agent/src/payloads.rs
Comment (ueno) [inline keylime-agent/src/payloads.rs:235]: I understand that the arguments of `worker` needs to be converted to an owned value before being sent to another thread, though it is a bit confusing to use owned value (`PathBuf`) and shared reference (`&Path`) in an intermixed manner. Maybe we could use `AsRef<Path>`?

```suggestion
    mount: impl AsRef<Path>,
```

File: keylime-agent/src/payloads.rs
Comment (ueno) [inline keylime-agent/src/payloads.rs:168]: Does this need to be `pub(crate)`, while it is only used in this file? Maybe good to review other functions as well.

File: keylime-agent/src/keys_handler.rs
Comment (aplanas) [inline keylime-agent/src/keys_handler.rs:201]: The length error is already in `e` if I read the code correctly. Maybe later there are other cases identified of why the conversion cannot be done.

File: keylime-agent/src/keys_handler.rs
Comment (aplanas) [inline keylime-agent/src/keys_handler.rs:219]: typo `encoding`

File: keylime-agent/src/keys_handler.rs
Comment (aplanas) [inline keylime-agent/src/keys_handler.rs:285]: too indented

File: keylime-agent/src/common.rs
Comment (ansasaki) [reply keylime-agent/src/common.rs]: Done! Thanks for the suggestion!

File: keylime-agent/src/common.rs
Comment (ansasaki) [reply keylime-agent/src/common.rs]: Done! Thank you!

File: keylime-agent/src/payloads.rs
Comment (ansasaki) [reply keylime-agent/src/payloads.rs]: Done! Thanks!

File: keylime-agent/src/payloads.rs
Comment (ansasaki) [reply keylime-agent/src/payloads.rs]: I reviewed the files I modified to make public only the functions that should be accessible externally.

File: keylime-agent/src/common.rs
Comment (ansasaki) [reply keylime-agent/src/common.rs]: Yes, modified to use named parameters for `format!`. Thank you!

File: keylime-agent/src/keys_handler.rs
Comment (ansasaki) [reply keylime-agent/src/keys_handler.rs:141]: Done!

File: keylime-agent/src/keys_handler.rs
Comment (ansasaki) [reply keylime-agent/src/keys_handler.rs]: Done!

File: keylime-agent/src/keys_handler.rs
Comment (ansasaki) [reply keylime-agent/src/keys_handler.rs]: I did not review all the cases, but modified some.

File: keylime-agent/src/keys_handler.rs
Comment (ansasaki) [reply keylime-agent/src/keys_handler.rs]: Thank you!

File: keylime-agent/src/keys_handler.rs
Comment (ansasaki) [reply keylime-agent/src/keys_handler.rs]: Thank you!

File: keylime-agent/src/common.rs
Comment (ansasaki) [reply keylime-agent/src/common.rs:184]: I think this is necessary to use the standard `std::result::Result` instead of the `crate::error::Result` which is in imported in the preamble.

**Comment by ansasaki** [comment]
It seems the change I introduced in e3ebc061d9f0bf27cb9d462c2ecd8b4c262e8979 makes `tarpaulin` to hang (see https://github.com/xd009642/tarpaulin/issues/1056).

I'll revert it for now.

**Comment by ansasaki** [comment]
@ueno @aplanas @lkatalin Could you please review this?

**Comment by aplanas** [comment]
> It seems the change I introduced in [e3ebc06](https://github.com/keylime/rust-keylime/commit/e3ebc061d9f0bf27cb9d462c2ecd8b4c262e8979) makes `tarpaulin` to hang (see [xd009642/tarpaulin#1056](https://github.com/xd009642/tarpaulin/issues/1056)).
> 
> I'll revert it for now.

I still see the error?

```
Jan 30 13:45:40.153 ERROR cargo_tarpaulin: Failed to get test coverage! Error: Failed to run tests: Error: Timed out waiting for test response
```

**Comment by ansasaki** [comment]
> > It seems the change I introduced in [e3ebc06](https://github.com/keylime/rust-keylime/commit/e3ebc061d9f0bf27cb9d462c2ecd8b4c262e8979) makes `tarpaulin` to hang (see [xd009642/tarpaulin#1056](https://github.com/xd009642/tarpaulin/issues/1056)).
> > I'll revert it for now.
> 
> I still see the error?
> 
> ```
> Jan 30 13:45:40.153 ERROR cargo_tarpaulin: Failed to get test coverage! Error: Failed to run tests: Error: Timed out waiting for test response
> ```

It seems this is back (sigh). It is difficult to debug this as I can't reproduce locally... I will investigate

File: keylime-agent/src/revocation.rs
Comment (aplanas) [inline keylime-agent/src/revocation.rs:201]: `.or_else(|| "".to_string())` ?

**Review by aplanas — APPROVED** [review]

File: keylime-agent/src/common.rs
Comment (ansasaki) [reply keylime-agent/src/common.rs:24]: The format you suggested is more compact, I'll replace.

File: keylime-agent/src/revocation.rs
Comment (ansasaki) [reply keylime-agent/src/revocation.rs:201]: Sorry, I do not get what you are suggesting here.

Comment (aplanas) [reply keylime-agent/src/revocation.rs:201]: Ah, maybe I am wrong: `actions = config_actions.unwrap_or_else(|| "".to_string())` ?

Comment (ansasaki) [reply keylime-agent/src/revocation.rs:201]: Ah, yes, that makes sense.

**Comment by ansasaki** [comment]
@aplanas I did not find the root cause that makes `tarpaulin` to hang. It definitely is something related with the instrumentation for test coverage, as it works fine when running the tests without it. There are multiple related issues open upstream on `tarpaulin` repo.

The only thing I could do was to update the image used for testing to a newer version (from Fedora 34 to Fedora 37).

If the issue makes the tests to fail frequently, let's disable the tests that runs external binaries or move the coverage measurement to a separate job which is not required to pass.

**Review by aplanas — APPROVED** [review]
For what I see LGTM

**Comment by ansasaki** [comment]
@ueno @aplanas @lkatalin Could you review/approve this?

**Comment by aplanas** [comment]
> @ueno @aplanas @lkatalin Could you please review/approve this?

AFAICT LGTM, yes

**Comment by ansasaki** [comment]
@THS-on Could you please review/approve this?

**Review by sergio-correia — APPROVED** [review]

File: keylime-agent/src/common.rs
Comment (THS-on) [inline keylime-agent/src/common.rs:48]: Why did this size change?

File: keylime-agent/src/main.rs
Comment (THS-on) [inline keylime-agent/src/main.rs:812]: Do we also need to shutdown the connection to the TPM cleanly (e.g. clean all sessions and flush the keys) or is that handled by the TPM library?

**Review by THS-on — APPROVED** [review]
General LGTM, but have not run it locally. 
I have not worked with much Rust sync code, so I cannot really comment on that part.

File: keylime-agent/src/common.rs
Comment (THS-on) [reply keylime-agent/src/common.rs:48]: Ah, yes that indeed makes way more sense.

File: keylime-agent/src/main.rs
Comment (ansasaki) [reply keylime-agent/src/main.rs:812]: From the messages I get when using `RUST_LOG=trace`, I believe the TPM library takes care of the context when the variable is dropped:
```
 INFO  tss_esapi::context                 > Closing context.
 INFO  tss_esapi::context                 > Flushing handle 1078035600
 INFO  tss_esapi::context                 > Flushing handle 1078035599
 INFO  tss_esapi::context                 > Closing handle 1078035594
 INFO  tss_esapi::context                 > Context closed.
```

**Comment by ansasaki** [comment]
@ashcrow @mbestavros @lukehinds @lkatalin @puiterwijk @mpeters @ueno

Could someone please review/approve this?

I believe we need to review/expand the reviewers who can approve changes to get merged. I'm trying to get this merged for weeks, but cannot get the required approval.

Also I would like to make changes to the repository (e.g. creating a template for issues, creating a template for PR checklist) but I do not have the required permission. Could someone with the right permissions please contact me?

**Comment by ashcrow** [comment]
I'm happy to approve given I see 3 approvals already and all outstanding comments have been resolved. My only caveat is that the Fedora tests failed. If this is a known issue then I'll add my approval, if not, let's get that fixed first :pray: 

@lkatalin Is there a discussion already on going about approvers by chance?

**Comment by ansasaki** [comment]
> I'm happy to approve given I see 3 approvals already and all outstanding comments have been resolved. My only caveat is that the Fedora tests failed. If this is a known issue then I'll add my approval, if not, let's get that fixed first pray

Done!

**Review by ashcrow — APPROVED** [review]
Review based on previous reviewers.

**Pull request merged**
